app/src/org/commcare/android/util/SigningUtil.java

```diff
@@ -0,0 +1,224 @@
+package org.commcare.android.util;
+
+import android.util.Pair;
+
+import org.commcare.android.logic.GlobalConstants;
+import org.spongycastle.jce.provider.BouncyCastleProvider;
+
+import java.io.BufferedReader;
+import java.io.IOException;
+import java.io.InputStreamReader;
+import java.net.URL;
+import java.security.InvalidKeyException;
+import java.security.KeyFactory;
+import java.security.NoSuchAlgorithmException;
+import java.security.PublicKey;
+import java.security.Signature;
+import java.security.SignatureException;
+import java.security.spec.X509EncodedKeySpec;
+
+/**
+ * A set of helper methods for verifying whether a message was genuinely sent from HQ. Currently we
+ * expcect the SMS in the format [COMMCARE APP - DO NOT DELETE] {BASE64-Encoded-Message} EG
+ *
+ * [commcare app - do not delete] Y2NhcHA6IGh0dHA6Ly9iaXQubHkvMU5JMUl6MyBzaWduYXR1cmU6IEvECygFUhiUH
+ * 3TRjC0lClQrpLR7lG//IpDYpRH7ComtZRjTirteXmPyM9fRgbPZ9K6jG9zEms9WQj55Uo7jTujKNYThjU8rJJmWLouJBr/Yn
+ * WobEupwzn6DP2FavPF1YLPbp0ZctOfymW3m4j3VZ0lR2dMOjmInMSBiInqICKid
+ *
+ * WHERE BASE64-Encoded-Message is decoded into:
+ *
+ * ccapp: <profile link> signature: <binary signature>
+ *
+ * And we can then verify that the profile link was in fact signed (using SHA256withRSA) by
+ * the CommCareHQ private key
+ *
+ * Created by wpride1 on 9/11/15.
+ */
+public class SigningUtil {
+
+    public static String retrieveParseVerifyMessage(String url) throws SignatureException, IOException {
+        String markedMessagePayload = convertUrlToPayload(url);
+        String messagePayload = trimMessagePayload(markedMessagePayload);
```

Comment:
you'll actually want to call `trimMessagePayload` first, since the String that's being passed as `url` to this method right now is actually:

`[commcare app - do not delete] <url>`